bwscanner/fetcher.py

```diff
@@ -82,9 +87,70 @@ def _getEndpoint(self, parsedURI, host=None, port=None):
             if hasattr(self, '_wrapContextFactory'):
                 tls_policy = self._wrapContextFactory(host, port)
             elif hasattr(self, '_policyForHTTPS'):
-                tls_policy = self._policyForHTTPS.creatorForNetloc(host, port)
+                tls_policy = self._policyForHTTPS().creatorForNetloc(host, port)
             else:
-                raise NotImplementedError(
-                    "can't figure out how to make a context factory")
+                raise NotImplementedError("Cannot create a TLS validation policy.")
             endpoint = self._tlsWrapper(tls_policy, endpoint)
         return endpoint
+
+
+class cancelableReadBodyProtocol(_ReadBodyProtocol):
+    def connectionLost(self, reason):
+        """
+        Deliver the accumulated response bytes to the waiting L{Deferred}, if
+        the response body has been completely received without error.
+
+        We can cancel the readBody Deferred after it is running. Canceling
+        the Deferred closes the connection. We want to check if the deferred
+        was already called to avoid raising an AlreadyCalled exception.
+        """
+        if reason.check(ResponseDone):
+            self.deferred.callback(b''.join(self.dataBuffer))
+        elif reason.check(PotentialDataLoss):
+            self.deferred.errback(
+                PartialDownloadError(self.status, self.message,
+                                     b''.join(self.dataBuffer)))
+        elif not self.deferred.called:
+            self.deferred.errback(reason)
+
+
+def cancelableReadBody(response):
+    """
+    Get the body of an L{IResponse} and return it as a byte string.
+
+    This is a helper function for clients that don't want to incrementally
+    receive the body of an HTTP response.
+
+    @param response: The HTTP response for which the body will be read.
+    @type response: L{IResponse} provider
+
+    @return: A L{Deferred} which will fire with the body of the response.
+        Cancelling it will close the connection to the server immediately.
+    """
+    def cancel(deferred):
+        """
+        Cancel a L{readBody} call, close the connection to the HTTP server
+        immediately, if it is still open.
+
+        @param deferred: The cancelled L{defer.Deferred}.
+        """
+        abort = getAbort()
+        if abort is not None:
+            abort()
+
+    d = defer.Deferred(cancel)
+    protocol = cancelableReadBodyProtocol(response.code, response.phrase, d)
+
+    def getAbort():
+        return getattr(protocol.transport, 'abortConnection', None)
+
```

Comment:
I'm not sure how `abortConnection` works here. `cancellableReadBody()` is the same as `twisted.web.client.readBody` but using the `cancelableReadBodyProtocol()` instead of the `readBodyProtocol`. 

The only difference is that the `cancelableReadBodyProtocol()` won't call the readBody deferred errback again. It appears that the `abortConnection` call does result in the connection being closed. Do you think we should change the `abortConnection` call? Or should we call `loseConnection()` if the protocol has no `abortConnection` method?